File: src/models/gemini.js
import { GoogleGenerativeAI } from '@google/generative-ai';
import { BaseModel } from './base.js';
import { config } from '../config.js';

export class GeminiModel extends BaseModel {
  constructor() {
    super('Gemini');
    this.genAI = new GoogleGenerativeAI(config.google.apiKey);
    this.model = this.genAI.getGenerativeModel({ model: 'gemini-2.5-pro' });
  }

  async sendMessage(messages, onChunk) {
    try {
      const startTimeMs = Date.now();
      const history = messages.slice(0, -1).map(msg => ({
        role: msg.role === 'assistant' ? 'model' : 'user',
        parts: [{ text: msg.content }]
      }));
      
      const chat = this.model.startChat({ history });
      const result = await chat.sendMessageStream(messages[messages.length - 1].content);
      
      let fullContent = '';
      let usage = null;
      
      for await (const chunk of result.stream) {
        const chunkText = chunk.text();
        if (chunkText) {
          fullContent += chunkText;
          if (onChunk) {
            onChunk(this.name, chunkText);
          }
        }
      }
      
      const response = await result.response;

      // Safely extract usage metadata
      if (response && response.usageMetadata) {
        try {
          const prompt_tokens = response.usageMetadata.promptTokenCount;
          const completion_tokens = response.usageMetadata.candidatesTokenCount;
          const total_tokens = response.usageMetadata.totalTokenCount;
          usage = {
            prompt_tokens,
            completion_tokens,
            total_tokens,
            // also provide input/output for generic handling
            input_tokens: prompt_tokens,
            output_tokens: completion_tokens
          };
        } catch (metadataError) {
          console.warn('Failed to extract usage metadata:', metadataError);
          usage = null;
        }
      }
      const responseTimeSeconds = (Date.now() - startTimeMs) / 1000;

      return {
        model: this.name,
        content: fullContent,
        usage,
        response_time: responseTimeSeconds
      };
    } catch (error) {
      return {
        model: this.name,
        error: error.message
      };
    }
  }
}

File: src_python/models/gemini.py
import google.generativeai as genai
from typing import List, Dict, Any, Optional, Callable
from .base import BaseModel
from ..config import config


class GeminiModel(BaseModel):
    def __init__(self):
        super().__init__('Gemini')
        genai.configure(api_key=config['google']['api_key'])
        self.model = genai.GenerativeModel('gemini-1.5-pro')
    
    async def send_message(
        self, 
        messages: List[Dict[str, str]], 
        on_chunk: Optional[Callable[[str, str], None]] = None
    ) -> Dict[str, Any]:
        try:
            formatted_messages = self.format_messages(messages)
            
            if on_chunk:
                chat = self.model.start_chat(history=formatted_messages[:-1])
                response = await chat.send_message_async(
                    formatted_messages[-1]['parts'][0],
                    stream=True
                )
                
                full_content = ''
                async for chunk in response:
                    if chunk.text:
                        full_content += chunk.text
                        on_chunk(self.name, chunk.text)
                
                return {
                    'model': self.name,
                    'content': full_content,
                    'usage': None
                }
            else:
                chat = self.model.start_chat(history=formatted_messages[:-1])
                response = await chat.send_message_async(
                    formatted_messages[-1]['parts'][0]
                )
                
                return {
                    'model': self.name,
                    'content': response.text,
                    'usage': {
                        'prompt_tokens': response.usage_metadata.prompt_token_count,
                        'completion_tokens': response.usage_metadata.candidates_token_count,
                        'total_tokens': response.usage_metadata.total_token_count
                    } if hasattr(response, 'usage_metadata') else None
                }
        except Exception as e:
            return {
                'model': self.name,
                'error': str(e)
            }
    
    def format_messages(self, messages: List[Dict[str, str]]) -> List[Dict[str, str]]:
        formatted = []
        for msg in messages:
            if msg['role'] == 'user':
                formatted.append({
                    'role': 'user',
                    'parts': [msg['content']]
                })
            elif msg['role'] == 'assistant':
                formatted.append({
                    'role': 'model',
                    'parts': [msg['content']]
                })
        return formatted